<img src="http://imgur.com/1ZcRyrc.png" style="float: left; margin: 20px; height: 55px">

## Lab: Cleaning Rock Song Data

_Authors: Dave Yerrington (SF)_

---


In [1]:
import pandas as pd
import numpy as np 
import seaborn as sns

%matplotlib inline

### 1. Load `rock.csv` and do an initial examination of its data columns.

In [14]:
rockfile = ("/Users/swllms/DAT-10-14/class material/Unit2/data/rock.csv")

In [17]:
# Load the data.
rf = pd.read_csv(rockfile)
rf.head()

Song Clean ARTIST CLEAN Release Year  \
0        Caught Up in You  .38 Special         1982   
1            Fantasy Girl  .38 Special          NaN   
2         Hold On Loosely  .38 Special         1981   
3  Rockin' Into the Night  .38 Special         1980   
4       Art For Arts Sake         10cc         1975   

                                COMBINED  First?  Year?  PlayCount  F*G  
0        Caught Up in You by .38 Special       1      1         82   82  
1            Fantasy Girl by .38 Special       1      0          3    0  
2         Hold On Loosely by .38 Special       1      1         85   85  
3  Rockin' Into the Night by .38 Special       1      1         18   18  
4              Art For Arts Sake by 10cc       1      1          1    1

In [16]:
# Look at the information regarding its columns.
rf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2230 entries, 0 to 2229
Data columns (total 8 columns):
Song Clean      2230 non-null object
ARTIST CLEAN    2230 non-null object
Release Year    1653 non-null object
COMBINED        2230 non-null object
First?          2230 non-null int64
Year?           2230 non-null int64
PlayCount       2230 non-null int64
F*G             2230 non-null int64
dtypes: int64(4), object(4)
memory usage: 139.5+ KB


### 2.  Clean up the column names.

Let's clean up the column names. There are two ways we can accomplish this:

#### 2.A Change the column names when you import the data using `pd.read_csv()`.

Notice that, when passing `names=[..A LIST OF STRING..]` with a number of columns that matches the number of strings in the passed list, you replace the column names.

NOTE: When you create custom column names, the first row of the `.csv` already represents a header. It is important to tell `pandas` to skip that row. The `skiprows=1` keyword argument to `read_csv()` will tell `pandas` to skip the first row.

In [18]:
# Change the column names when loading the '.csv':
column_names = ['Song', 'Artist', 'RYR', 'Song_Artist', 'First', 'Year', 'CountPlay', 'FG']
rf = pd.read_csv(rockfile, names=column_names, skiprows =1)

In [20]:
rf.head()

Song       Artist   RYR  \
0        Caught Up in You  .38 Special  1982   
1            Fantasy Girl  .38 Special   NaN   
2         Hold On Loosely  .38 Special  1981   
3  Rockin' Into the Night  .38 Special  1980   
4       Art For Arts Sake         10cc  1975   

                             Song_Artist  First  Year  CountPlay  FG  
0        Caught Up in You by .38 Special      1     1         82  82  
1            Fantasy Girl by .38 Special      1     0          3   0  
2         Hold On Loosely by .38 Special      1     1         85  85  
3  Rockin' Into the Night by .38 Special      1     1         18  18  
4              Art For Arts Sake by 10cc      1     1          1   1

#### 2.B Change column names using the `.rename()` function.

The `.rename()` function takes an argument, `columns=name_dict`, in which `name_dict` is a dictionary containing the original column names as keys and the new column names as values.

In [25]:
# Change the column names using the `.rename()` function.
name_dict = {'Song': 'Songs','Artist': 'ArtistC','RYR': 'Rel_Yr','Song_Artist': 'Song_Art','First': '1st','Year': 'YR', 'PlayCount': 'PlayC', 
'FG': 'F*G'}
rf.rename(columns=name_dict) 

### can also use rf = rf.rename({}) axis=1, inplace=True) 
### new cols = [col.lower() for col in rf.colums.tolist]

Songs        ArtistC Rel_Yr  \
0                                Caught Up in You    .38 Special   1982   
1                                    Fantasy Girl    .38 Special    NaN   
2                                 Hold On Loosely    .38 Special   1981   
3                          Rockin' Into the Night    .38 Special   1980   
4                               Art For Arts Sake           10cc   1975   
5                                      Kryptonite   3 Doors Down   2000   
6                                           Loser   3 Doors Down   2000   
7                                   When I'm Gone   3 Doors Down   2002   
8                                      What's Up?  4 Non Blondes   1992   
9                                      Take On Me           a-ha   1985   
10                          Baby, Please Don't Go          AC/DC    NaN   
11                                  Back In Black          AC/DC   1980   
12                                        Big Gun          AC/DC   1993   
13                         CAN'T STOP ROCK'N'ROLL          AC/DC    NaN   
14                    Dirty Deeds Done Dirt Cheap          AC/DC   1976   
15                        For Those About To Rock          AC/DC   1981   
16                               Girls Got Rhythm          AC/DC    NaN   
17                                 Hard As A Rock          AC/DC   1995   
18                             Have a Drink On Me          AC/DC   1980   
19                                    Hells Bells          AC/DC   1980   
20                                Highway To Hell          AC/DC   1979   
21                     It's A Long Way To The Top          AC/DC   1975   
22                                      Jailbreak          AC/DC   1984   
23                              Let There Be Rock          AC/DC   1977   
24                                Let's Get It Up          AC/DC    NaN   
25                                      Live Wire          AC/DC    NaN   
26                                     Moneytalks          AC/DC    NaN   
27                                  Night Prowler          AC/DC   1979   
28            Rock and Roll Ain't Noise Pollution          AC/DC   1980   
29                                Shoot To Thrill          AC/DC    NaN   
...                                           ...            ...    ...   
2200                      Owner of a Lonely Heart            Yes   1983   
2201                                   Roundabout            Yes   1971   
2202                             Starship Trooper            Yes   1971   
2203                            Wonderous Stories            Yes    NaN   
2204                         Yours Is No Disgrace            Yes   1971   
2205                        Tell Me What You Want          Zebra    NaN   
2206                        Who's Behind The Door          Zebra    NaN   
2207                           Time of the Season        Zombies   1968   
2208                    A Fool for Your Stockings         ZZ Top    NaN   
2209             Arrested For Driving While Blind         ZZ Top   1976   
2210                 Beer Drinkers & Hell Raisers         ZZ Top   1973   
2211                             Cheap Sunglasses         ZZ Top   1980   
2212                         Gimme All Your Lovin         ZZ Top   1983   
2213                        Got Me Under Pressure         ZZ Top   1983   
2214                            Heard It On the X         ZZ Top   1975   
2215                                  I Thank You         ZZ Top    NaN   
2216                      I'm Bad, I'm Nationwide         ZZ Top    NaN   
2217                      Jesus Just Left Chicago         ZZ Top   1973   
2218                                Just Got Paid         ZZ Top    NaN   
2219                                    La Grange         ZZ Top   1973   
2220                                         Legs         ZZ Top   1983   
2221                     My Head's In Mississippi         ZZ Top    N

### 3. Subsetting data where null values exist.

We have mixed `str` and `NaN` values in the `release` column. `NaN` stands for "not a number" and is the way `pandas` handles "nulls" or nonexistent data. We can use the `.isnull()` method of a Series to find null values.

Print the header of the data subset to where the `release` column is null values.

In [28]:
# Show records where df['release'] is null
rf = pd.read_csv(rockfile)
rf.head()

Song Clean ARTIST CLEAN Release Year  \
0        Caught Up in You  .38 Special         1982   
1            Fantasy Girl  .38 Special          NaN   
2         Hold On Loosely  .38 Special         1981   
3  Rockin' Into the Night  .38 Special         1980   
4       Art For Arts Sake         10cc         1975   

                                COMBINED  First?  Year?  PlayCount  F*G  
0        Caught Up in You by .38 Special       1      1         82   82  
1            Fantasy Girl by .38 Special       1      0          3    0  
2         Hold On Loosely by .38 Special       1      1         85   85  
3  Rockin' Into the Night by .38 Special       1      1         18   18  
4              Art For Arts Sake by 10cc       1      1          1    1

In [30]:
rf[rf['Release Year'].isnull()]

Song Clean              ARTIST CLEAN  \
1                           Fantasy Girl               .38 Special   
10                 Baby, Please Don't Go                     AC/DC   
13                CAN'T STOP ROCK'N'ROLL                     AC/DC   
16                      Girls Got Rhythm                     AC/DC   
24                       Let's Get It Up                     AC/DC   
25                             Live Wire                     AC/DC   
26                            Moneytalks                     AC/DC   
29                       Shoot To Thrill                     AC/DC   
31                              Sin City                     AC/DC   
35        What Do You Do For Money Honey                     AC/DC   
42                 Baby, Please Don't Go                 Aerosmith   
45                         Come Together                 Aerosmith   
53                      Kings and Queens                 Aerosmith   
60                        One Way Street                 Aerosmith   
66                        The Other Side                 Aerosmith   
71                     You See Me Crying                 Aerosmith   
75                               Fantasy                 Aldo Nova   
76                           BE MY LOVER              Alice Cooper   
80                         House Of Fire              Alice Cooper   
100                          One Way Out      Allman Brothers Band   
102                           Southbound      Allman Brothers Band   
104                        Whippin' Post      Allman Brothers Band   
110                       Immigrant Song                Ann Wilson   
111                               Roller                April Wine   
113              Coming into Los Angeles              Arlo Guthrie   
114                              Wayside             Artful Dodger   
118                               Spooky    Atlanta Rhythm Section   
120                Blondes In Black Cars                 Autograph   
122   Rock And Roll Party In The Streets                       Axe   
124             Roll On Down The Highway  Bachman-Turner Overdrive   
...                                  ...                       ...   
2149              Take Your Whiskey Home                 Van Halen   
2150                           Top Jimmy                 Van Halen   
2151                    Top of the World                 Van Halen   
2157                    And It Stoned Me              Van Morrison   
2158                          BLUE MONEY              Van Morrison   
2162                              Gloria              Van Morrison   
2164                  Jackie Wilson Said              Van Morrison   
2168              Edge of a Broken Heart                     Vixen   
2174                   Uncle Tom's Cabin                   Warrant   
2178              Street Corner Serenade                Wet Willie   
2179                                Wait                White Lion   
2184                       Slow An' Easy                Whitesnake   
2185                  Still of the Night                Whitesnake   
2186                        Who Are You?                   Why Not   
2188                         Night Moves                  Wildlife   
2189                   We've Got Tonight                  Wildlife   
2190                    Can't Get Enough                    Winger   
2191             Headed For A Heartbreak                    Winger   
2192                           Seventeen                    Winger   
2194                    Summertime Girls                       Y&T   
2203                   Wonderous Stories                       Yes   
2205               Tell Me What You Want                     Zebra   
2206               Who's Behind The Door                     Zebra   
2208           A Fool for Your Stockings                    ZZ Top   
2215                         I Thank You                    ZZ Top   
2216             I'm Bad, I'm Nationwide

### 4. Update slices of your DataFrame based on mask selection/slices.

In many scenarios, we want to upate values in our DataFrame according to criteria. Let's say we wanted to set all of the null values in `release` to 0.

With newer versions of `pandas`, in order to manipulate data in the original DataFrame, we have to use `.loc` while performing reassignment using a mask and an index.

For example, the following won't always work:
```python
df[row_mask]['column_name'] = new_value
```

The best way to accomplish the same task is:
```python
df.loc[row_mask, 'column_name'] = new_value
```

For multiple column assignment, you would use:
```python
df.loc[row_mask, ['col_1', 'col_2', 'col_3']] = new_value
```

#### 4.A Let's try it out. Make all of the null values in `release` 0.

In [83]:
# Replace release nulls with 0
###rf.loc[:, rf.isnull().sum() > 0 ]This shows all the Null Values

rel_null_val = rf['Release Year'].isnull()
rf.loc[rel_null_val, 'Release Year'] = 0

rf.head()

Song Clean ARTIST CLEAN Release Year  \
0        Caught Up in You  .38 Special         1982   
1            Fantasy Girl  .38 Special            0   
2         Hold On Loosely  .38 Special         1981   
3  Rockin' Into the Night  .38 Special         1980   
4       Art For Arts Sake         10cc         1975   

                                COMBINED  First?  Year?  PlayCount  F*G  
0        Caught Up in You by .38 Special       1      1         82   82  
1            Fantasy Girl by .38 Special       1      0          3    0  
2         Hold On Loosely by .38 Special       1      1         85   85  
3  Rockin' Into the Night by .38 Special       1      1         18   18  
4              Art For Arts Sake by 10cc       1      1          1    1

#### 4.B Verify that `release` contains no null values.

In [49]:
# A:
rf.isnull()

Song Clean  ARTIST CLEAN  Release Year  COMBINED  First?  Year?  \
0          False         False         False     False   False  False   
1          False         False         False     False   False  False   
2          False         False         False     False   False  False   
3          False         False         False     False   False  False   
4          False         False         False     False   False  False   
5          False         False         False     False   False  False   
6          False         False         False     False   False  False   
7          False         False         False     False   False  False   
8          False         False         False     False   False  False   
9          False         False         False     False   False  False   
10         False         False         False     False   False  False   
11         False         False         False     False   False  False   
12         False         False         False     False   False  False   
13         False         False         False     False   False  False   
14         False         False         False     False   False  False   
15         False         False         False     False   False  False   
16         False         False         False     False   False  False   
17         False         False         False     False   False  False   
18         False         False         False     False   False  False   
19         False         False         False     False   False  False   
20         False         False         False     False   False  False   
21         False         False         False     False   False  False   
22         False         False         False     False   False  False   
23         False         False         False     False   False  False   
24         False         False         False     False   False  False   
25         False         False         False     False   False  False   
26         False         False         False     False   False  False   
27         False         False         False     False   False  False   
28         False         False         False     False   False  False   
29         False         False         False     False   False  False   
...          ...           ...           ...       ...     ...    ...   
2200       False         False         False     False   False  False   
2201       False         False         False     False   False  False   
2202       False         False         False     False   False  False   
2203       False         False         False     False   False  False   
2204       False         False         False     False   False  False   
2205       False         False         False     False   False  False   
2206       False         False         False     False   False  False   
2207       False         False         False     False   False  False   
2208       False         False         False     False   False  False   
2209       False         False         False     False   False  False   
2210       False         False         False     False   False  False   
2211       False         False         False     False   False  False   
2212       False         False         False     False   False  False   
2213       False         False         False     False   False  False   
2214       False         False         False     False   False  False   
2215       False         False         False     False   False  False   
2216       False         False         False     False   False  False   
2217       False         False         False     False   False  False   
2218       False         False         False     False   False  False   
2219       False         False         False     False   False  False   
2220       False         False         False     False   False  False   
2221       False         False         False     False   False  False   
2222       False         False         False     False   False  

In [70]:
rf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2230 entries, 0 to 2229
Data columns (total 8 columns):
Song Clean      2230 non-null object
ARTIST CLEAN    2230 non-null object
Release Year    2230 non-null object
COMBINED        2230 non-null object
First?          2230 non-null int64
Year?           2230 non-null int64
PlayCount       2230 non-null int64
F*G             2230 non-null int64
dtypes: int64(4), object(4)
memory usage: 139.5+ KB


### 5. Ensure that the data types of the columns make sense. 

Verifying column data types is a critical part of data munging. If columns have the wrong data type, then there is usually corrupted or incorrect data in some of the observations.

#### 5.A Look at the data types for the columns. Are any incorrect given what the data represents?

In [52]:
# A:
rf.dtypes ###Object are genric, sim to a python string, example it could be a list.

Song Clean      object
ARTIST CLEAN    object
Release Year    object
COMBINED        object
First?           int64
Year?            int64
PlayCount        int64
F*G              int64
dtype: object

### 6. Investigate and clean up the `release` column.

The `release` column is a string data type when it should be an integer.

#### 6.A Figure out what value(s) are causing the `release` column to be encoded as a string instead of an integer.

In [71]:
# A:
rf['Release Year'].unique()

array(['1982', 0, '1981', '1980', '1975', '2000', '2002', '1992', '1985',
       '1993', '1976', '1995', '1979', '1984', '1977', '1990', '1986',
       '1974', '2014', '1987', '1973', '2001', '1989', '1997', '1971',
       '1972', '1994', '1970', '1966', '1965', '1983', '1955', '1978',
       '1969', '1999', '1968', '1988', '1962', '2007', '1967', '1958',
       '1071', '1996', '1991', '2005', '2011', '2004', '2012', '2003',
       '1998', '2008', '1964', '2013', '2006', 'SONGFACTS.COM', '1963',
       '1961'], dtype=object)

In [80]:
rf['Release Year'] = rf['Release Year'].astype(str) ###Another way to look up is: 

In [82]:
rf[~rf['Release Year'].str.isdigit()] ###~ means not true

Empty DataFrame
Columns: [Song Clean, ARTIST CLEAN, Release Year, COMBINED, First?, Year?, PlayCount, F*G]
Index: []

#### 6.B Look at the rows in which there is incorrect data in the `release` column.

In [72]:
# A:
obj_rel = rf['Release Year'] == 'SONGFACTS.COM'
rf[obj_rel]

Song Clean    ARTIST CLEAN   Release Year  \
1504  Bullfrog Blues  Rory Gallagher  SONGFACTS.COM   

                              COMBINED  First?  Year?  PlayCount  F*G  
1504  Bullfrog Blues by Rory Gallagher       1      1          1    1

#### 6.C. Clean up the data. Normally we may replace the offending data with null np.nan values, however we previously converted all of the nan values in the release column to zeros so we might as well continue with the same practice. Replacing with 0 (or nan) will allow us to convert the column to numeric.

In [85]:
rf.loc[obj_rel, 'Release Year'] = np.nan ###np.nan is what? 
rf['Release Year'] = rf['Release Year'].map(lambda x: float(x))

rf.dtypes ###NOT sure how this happened. 

Song Clean       object
ARTIST CLEAN     object
Release Year    float64
COMBINED         object
First?            int64
Year?             int64
PlayCount         int64
F*G               int64
dtype: object

### 7. Get summary statistics for the `release` column using the `.describe()` function.

Now that the `release` column is finally a numeric data type, we can apply the `.describe()` function.  

#### 7.A Print out the summary stats for the `release` column. What is the earliest and latest release date?

In [86]:
# A:
rf['Release Year'].describe()

count    2230.000000
mean     1465.331390
std       867.196161
min         0.000000
25%         0.000000
50%      1973.000000
75%      1981.000000
max      2014.000000
Name: Release Year, dtype: float64

#### 7.B Based on the summary statistics, is there anything else wrong with the `release` column? 

In [16]:
# A:


_Looking at the DataFrame that contains the year 1071, we can see that the year was probably corrupted and should be replaced with something else if possible._

### 8. Make changes and investigate using custom functions with `.apply()`.

Let's say we want to traverse every single row in our data set and apply a function to that row.

#### 8.A Write a function that will take a row of a DataFrame and print out the song, artist, and whether or not the release date is < 1970.


In [17]:
# A:
def release_inspector(row)

#### 8.B Using the `.apply()` function, apply the function you wrote to the first four rows of the DataFrame.

You will need to tell the `apply` function to operate row by row. Setting the keyword argument as `axis=1` indicates that the function should be applied to each row individually.

In [18]:
# A: